dmoj/executors/mixins.py

```diff
@@ -12,38 +12,76 @@
 from dmoj.utils import setbufsize_path
 from dmoj.utils.unicode import utf8bytes
 
-BASE_FILESYSTEM = [
-    '/dev/(?:null|tty|zero|u?random)$',
-    '/usr/(?!home)',
-    '/lib(?:32|64)?/',
-    '/opt/',
-    '/etc$',
-    '/etc/(?:localtime|timezone)$',
-    '/usr$',
-    '/tmp$',
-    '/$',
-]
-BASE_WRITE_FILESYSTEM = ['/dev/stdout$', '/dev/stderr$', '/dev/null$']
+from .filesystem_rules import FilesystemPolicy, Dir, File
+
+BASE_FILESYSTEM = FilesystemPolicy(
+    File("/dev/null"),
+    File("/dev/tty"),
+    File("/dev/zero"),
+    File("/dev/urandom"),
+    File("/dev/random"),
+    *[ Dir(f"/usr/{d}") for d in os.listdir("/usr") if d != "home" ],
+
+    Dir("/lib"),
+    Dir("/lib32"),
+    Dir("/lib64"),
+    Dir("/opt"),
+
+    File("/etc"),
+    File("/etc/localtime"),
```

Comment:
But, either way this isn't actually consistent. Either `Dir('/etc')` means:

- "allow everything under `/etc`, recursively", which is scary, and it shouldn't, but in which case the `/etc/localtime` rules are duplicative, or
- it means "allow only /etc, recursively", in which case the `/etc/localtime` rules make sense, but `Dir('/opt')` above is not enough.

Reply:
The rules outlined in https://github.com/DMOJ/judge-server/issues/871 distinguish between these cases with `Self()` and `All()` atoms:

```python
    Dir("etc", (
        Self(),
        File("localtime"),
        File("timezone"),
    )),
    ...
    Dir("opt", (All(),))
```